JS/Lesson19/saffet/exercises.js

```diff
@@ -1,8 +1,201 @@
+        this.make = make;
+        this.model = model;
+        this.year = year;
+    }
+
+    start(){
+        console.log(`the vehicle make ${this.make}, model ${this.model} and year ${this.year} has started`);
+    }
+}
+
+class Car extends Vehicle{
+    constructor(make, model, year, numDoors){
+        super(make, model, year);
+        this.numDoors = numDoors;
+    }
+
+    start(){
+        console.log(`the car ${this.make} ${this.model} ${this.year} and number of door is ${this.numDoors} has started`);
+    }
+}
+
+const vehicleInstance = new Vehicle("Audi", "RS6", 2019);
+const carInstance = new Car("Nissan", "GTR", "2017", 2);
+
+vehicleInstance.start();
+carInstance.start();
+
+console.log("==================================");
+// 4. Create a BankAccount class with balance and interestRate properties, and deposit() and withdraw() methods. 
+// Add a transactionHistory property that is an array of objects representing each transaction made on the account, 
+// with properties for type (either "deposit" or "withdrawal"), amount, and date ({type: "withdrawal", amount: amount, 
+// date: new Date().toLocaleDateString()}). Implement deposit() and withdraw() methods that add new transactions to the 
+// transactionHistory array. Implement a get transactionHistory() getter method that returns a copy of the transaction history array, 
+// and a get currentBalance() getter method that calculates and returns the current balance of the account based on the 
+// transaction history.
+console.log("Exercise 4: ");
+
+class BankAccount{
+    constructor(balance, interestRate){
+        this. balance = balance;
+        this.interestRate = interestRate;
+        this._transactionHistory = [];
+    }
+
+    deposit(amount){
+        if(amount>0){
+            this.balance += amount;
+            const transaction = {
+                type: "deposit",
+                amount: amount,
+                date: new Date().toLocaleDateString()
+            };
+            this._transactionHistory.push(transaction);
+            console.log(`Succesfully deposited ${amount} into the bank account`);
+        }
+        else{
+            console.log("Please provide a positive value.");
+        }
+    }
+
+    withdraw(amount){
+        if(amount>0){
+            if(this.balance >= amount){
+                this.balance -= amount;
+                const transaction = {
+                    type: "Widthrawal",
+                    amount : amount,
+                    date: new Date().toLocaleDateString()
+                }
+                this._transactionHistory.push(transaction);
+                console.log(`Successfully witdhrew ${amount} from the account.`);
+            }
+            else{
+                console.log("Insufficient balance.");
+            }
+        }
+        else{
+            console.log("Provide a positive value.");
+        }
+    }
+
+    get transactionHistory() {
+        return [...this._transactionHistory];
```

Comment:
Nice that you return a shallow copy of the data instead of the original, it's always good idea to protect the original values :)